File: src/main/scala/2017/day-10.scala
package adventofcode2017

object Day10 {
  lazy val input: String = "227,169,3,166,246,201,0,47,1,255,2,254,96,3,97,144"

  def sparseHash(nums: List[Int], times: Int, listLength: Int): List[Int] = {
    val list = (0 until listLength).toList ++ (0 until listLength).toList
    val (res, _, _) = (0 until times).foldLeft((list, 0, 0)) {
      case ((list, pos, skip), _) => 
        val (newList, newPos, newSkip) = nums.foldLeft((list, pos, skip)) {
          case ((list, pos, skip), n) => 
            val part = list.slice(pos, pos + n).reverse
            val loopI = part.length - Math.max(pos + n - listLength, 0);
            val partToStart = part.slice(loopI, part.length);
            val half = partToStart ++
              list.slice(partToStart.length, pos) ++
              part.slice(0, loopI) ++
              list.slice(pos+n, list.length)
              val nextList = half.slice(0, listLength) ++ half.slice(0, listLength)
            (nextList, (pos + n + skip) % listLength, skip+1)
        }
        (newList, newPos, newSkip)
      
    }
    res.slice(0, listLength)
  }

  def denseHash(nums: List[Int]): List[Int] = {
    val res = (0 until 16).foldLeft(List[Int]()) { (x, i) => 
      val value = nums.slice(i*16, i*16+16).foldLeft(0)(_^_)
      x :+ value
    }
    res
  }

  def main(args: Array[String]): Unit = {
    val nums = input.toCharArray.map(_.toInt).toList ++ List(17, 31, 73, 47, 23)
    val sparseHashVal = sparseHash(nums, 64, 256)
    val denseHashVal = denseHash(sparseHashVal);
    val knotHash = denseHashVal.foldLeft("") { (res, n) => 
      res + (s"0${n.toHexString}" takeRight 2)
    }
    println(knotHash)
  }
}

File: 2017/js/day-10.js
function reverse(arr) {
  return arr.reduce((res, n) => {
    res.unshift(n);
    return res;
  }, []);
}

function sparseHasStep(listLength, [list, pos, skip], n) {
  const part = reverse(list.slice(pos, pos + n));
  const loopI = part.length - Math.max(pos + n - listLength, 0);
  const partToStart = part.slice(loopI);
  let half = partToStart
    .concat(list.slice(partToStart.length, pos))
    .concat(part.slice(0, loopI))
    .concat(list.slice(pos + n));
  const newList = [...half.slice(0, listLength), ...half.slice(0, listLength)];
  return [newList, (pos + n + skip) % listLength, skip + 1];
}

function sparseHash(nums, times, listLength) {
  let list = [];
  for (let i = 0; i < listLength; i++) list.push(i);
  list = [...list.slice(0, listLength), ...list.slice(0, listLength)];

  let pos = 0;
  let skip = 0;
  for (let i = 0; i < times; i++) {
    [list, pos, skip] = nums.reduce(([list, pos, skip], n) =>
      sparseHasStep(listLength, [list, pos, skip], n),
      [list, pos, skip]);
  }
  return list.slice(0, listLength);
}

function denseHash(nums) {
  let result = [];
  for (let i = 0; i < 16; i++) {
    const start = i * 16;
    result.push(nums.slice(start, start + 16).reduce((res, n) => res ^ n, 0));
  }
  return result;
}

let input = '227,169,3,166,246,201,0,47,1,255,2,254,96,3,97,144';
const nums = input.replace(/\s/g, '').split('').map(x => x.charCodeAt(0)).concat([17, 31, 73, 47, 23]);
const sparseHashVal = sparseHash(nums, 64, 256);
const denseHashVal = denseHash(sparseHashVal);
const knotHash = denseHashVal.reduce((res, n) => {
  const hex = n.toString(16);
  return res + (hex.length === 1 ? `0${hex}` : hex);
}, '');
console.log(knotHash);
